### ISSUE TITLE
About monodepth2 result in the paper

### ISSUE THREAD

**Issue by myalos** [SJTU-ViSYS/StructDepth#9]
Thanks for sharing the code !
I noticed that in the table1 of structdepth paper, the evaluation on NYUv2 of monodepth2 is (delta1 77.1 delta2 94.8 delta3 98.7), which is different from table1 of P2Net paper which is (delta1 0.748, delta2 0.942, delta3 0.986)
And I tried to train monodepth2 on the NYUv2 dataset, only change the opt.max_depth from 100.0 to 10.0 and opt.scales from [0, 1, 2, 3] to [0], other parameters remaining the same, and get the evalution result (delta1 0.691 delta2 0.911 delta3 0.973.)
So can you share the recipe of training monodepth2 on the NYUv2 dataset as you do in your paper to help me reproduce the result and save electricity.
Thanks in advance.

**Comment by myalos** [comment]
I changed the epoch from 20 to 41 and StepLR from 15 0.1 to MultiStepLR [26, 36] 0.1 as StructDepth do, and get the result
abs_rel |   sq_rel |     rmse | rmse_log |     lg10 |       a1 |       a2 |       a3 | 
&   0.165  &   0.137  &   0.604  &   0.206  &   0.069  &   0.762  &   0.945  &   0.986  \\
I think changing the scale from [0] to [0, 1, 2, 3] will get the result (delta1 77.1 delta2 94.8 delta3 98.7) in the paper

**Comment by Cresynia** [comment]
Can I ask you how to train monodepth2 on the NYUv2 dataset,I'm a beginner and met the problems,hope you can give some help,thanks!

**Comment by myalos** [comment]
Copy the nyu_dataset.py to monodepth2 dataset folder, remove vps and dso related code, change the frames_idx from 5 frames to 3 frames, and change the code in trainer.py, options.py, dataset/__init__.py correspondingly，hope it helps.

**Comment by Cresynia** [comment]
Thanks! I'll try it.

**Comment by Cresynia** [comment]
Hi! I got some questions want to ask.
When I finished training a model,I want to visualized it on the NYUv2 dataset,and I used the python inference_single_image.py to do it,but I don't know whether it's good or not, how can I get the groudtruth depth map?
And I get confused about the parameter it set,like λ1，λ2，λ3，w=0.85，α，β，I don't know why.
And I used the pretrain model,batchsize32,epoch50,get the result:RMSE 0.543 AbsRel 0.143 Log10 0.061 δ1 81.1 δ2 95.4 δ3 98.8,have a little difference from the paper,I don't know if I made errors or it's normal.
And I discoverd the paper use the post precessing[19] have a better result,want to ask what it is.
Thanks!

**Comment by myalos** [comment]
Q: how can I get the groudtruth depth map? 
A: I remeber groudtruth depth map is in the test set (h5 file, depth key)
Q: ... I don't know why ?
A: These parameters are hyperparameters. Maybe these parameters are determined by some search method like random search(search space is determined empirically.
Q: I don't know if I made errors or it's normal.
A: The author provide two pretrained model, you should try another one.
Q: about post precessing
A:  This method origins from the monodepth paper, and it can help the model to get better result.
Hope this help

**Comment by Cresynia** [comment]
Thanks! And could you explain the h5 file,I don't really understand. Does it means I can get the depth value from the h5 file,and use it to visualize?

**Comment by myalos** [comment]
h5 file is something like zip file.
Q: How to get the depth value from the h5 file? 
A: you can refer to the code in datasets/nyu_dataset.py h5_loader function.
Q: How to visualize ?
A: I visualize the depth map using cv2.applyColorMap function, you can refer to the code in 
 inference_single_image.py.

**Comment by Cresynia** [comment]
Ok,I'll then have a try,thanks!



---Original---
From: ***@***.***&gt;
Date: Tue, Mar 21, 2023 11:59 AM
To: ***@***.***&gt;;
Cc: ***@***.******@***.***&gt;;
Subject: Re: [SJTU-ViSYS/StructDepth] About monodepth2 result in the paper(Issue #9)




 
h5 file is something like zip file.
 Q: How to get the depth value from the h5 file?
 A: you can refer to the code in datasets/nyu_dataset.py h5_loader function.
 Q: How to visualize ?
 A: I visualize the depth map using cv2.applyColorMap function, you can refer to the code in
 inference_single_image.py.

**Comment by Cresynia** [comment]
Hi! When I learn the extract_vps_nyu.py, I met some problems.I don't know why they choose focal_length = fx,not fy and how to choose the seed.And how the length_thresh influence the result?I do it on a single image,but I still can't understand.And why they do np.vstack([vps, -vps]),not just vps?

**Comment by myalos** [comment]
Q: I don't know why they choose focal_length?
A: I think either choose fx or choose fy is ok.
Q: how to choose the seed ?
A: seed is one of hyperparameters. As far as i know, there is no guideline to choose good seed.
Q: length_thresh influence the result?
A: length_thresh lower more line is chosen and noise is higher, but if length_thresh is too high, then maybe the algorithm ouputs zero vanishing point, you can look the paper of this vanishing point detection method.
Q: why they do np.vstack([vps, -vps])?
A: since they will choose the main direction for each pixel in the photo, main direction should account for -vps, you can draw 
 on paper and think about it.
Hope this help

**Comment by Cresynia** [comment]
Oh！Thanks！

**Comment by Cresynia** [comment]
Hi!When I learn the monodepth2 codes, I met some problems. In the compute_losses function of trainer.py, it use 
            if not self.opt.disable_automasking:
                outputs["identity_selection/{}".format(scale)] = (
                    idxs > identity_reprojection_loss.shape[1] - 1).float()
to do the auto-masking,I think this represent μ in the paper,(I don't know whether I think right)but I don't find when they use it then.
They just to add the mean loss of it to the total loss, don't use the μ.

**Comment by myalos** [comment]
The code you post here is used for visualization. Variable "idxs" is the index of minimum loss in variable "combined", and idxs > identity_reprojection_loss.shape[1] - 1 means the total loss choose reprojection loss rather than identity loss.
μ in the paper corresponds to variable "to_optimise", 
and it is used in the code "loss += to_optimise.mean()".
Hope this help.

**Comment by Cresynia** [comment]
Oh! I get it!
Maybe like this?
to_optimise, idxs = torch.min(combined, dim=1)
to_optimise get the minimum loss of all identity_reprojection_loss and reprojection loss.If the minimum loss belongs to reprojection loss,represent μ=1,and add it to the total loss.If minimum loss belongs to identity_reprojection_loss, in this case,it can illustrate that these pixles are nearly have  same velocity of camera and need to be filtered out,and minimum loss is nearly 0 because they nearly have  same velocity,so can add it to the loss directly(represent μ=0).

**Comment by myalos** [comment]
yes，“μ in the paper corresponds to variable "to_optimise"“ is wrong, μ in the paper is a boolean value. "idx" corresponds to μ. “I don't find when they use it then.” I think to_optimise is the answer, but describe these clearly in english is not trivial for me. QAQ. Hope you can understand my expression.

**Comment by myalos** [comment]
yes

**Comment by Cresynia** [comment]
OK，thanks！



---Original---
From: ***@***.***&gt;
Date: Sat, Apr 1, 2023 12:34 PM
To: ***@***.***&gt;;
Cc: ***@***.******@***.***&gt;;
Subject: Re: [SJTU-ViSYS/StructDepth] About monodepth2 result in the paper(Issue #9)




 
yes
